### PR TITLE
KAFKA-3517: Add documentation for SASL/PLAIN

### PR DESCRIPTION
**Pull request by rajinisivaram**
Documentation corresponding to KIP-43 - SASL/PLAIN and multiple mechanism support.

### CODE REVIEW

**Comment by rajinisivaram** [comment]
The code changes corresponding to this doc are under KAFKA-3149 (not yet committed).

File: docs/security.html [hunk 1]
```diff
@@ -257,93 +321,191 @@
...
+          <li>Add a suitably modified JAAS file similar to the one below to each Kafka broker's config directory, let's call it kafka_server_jaas.conf for this example:
+            <pre>
+    KafkaServer {
+        org.apache.kafka.common.security.plain.PlainLoginModule required
+        username="admin"
+        password="admin-secret"
+        user_admin="admin-secret"
+        user_alice="alice-secret";
+    };</pre></li>
+          This configuration defines two users (<i>admin</i> and <i>alice</i>). <i>admin</i> is used for inter-broker communication.
```
Comment (edenhill) [inline docs/security.html]: Maybe explicitly mention that alice is a client user?
Is user_N defining a client user? If so, does this mean that admin may also connect as a client in the above example?
I'm guessing username/password is for outgoing sessions, while user_N is for incoming?

Comment (rajinisivaram) [inline docs/security.html]: @edenhill Thank you for the review. `username` and `password` define the username/password combination used by client connections. Kafka clients use the login context `KafkaClient` and brokers use the login context `KafkaServer`. So in the example, the client connections use the username `alice` and the inter-broker connections use the username `admin`. In KafkaServer, `user_N=M` is the server-side configuration defining all the valid username/password pairs that the server validates against. In this example, any client may connect using username `admin` by setting `username=admin` in the `KafkaClient` context. Hope that helps. I will update the docs to clarify these.

File: docs/security.html [hunk 2]
```diff
@@ -182,22 +183,95 @@
...
+        <li>ZooKeeper uses "zookeeper" as the service name by default. If you
+          want to change this, set the system property
+          <tt>zookeeper.sasl.client.username</tt> to the appropriate name
+          (<i>e.g.</i>, <tt>-Dzookeeper.sasl.client.username=zk</tt>).</li>
+      </ol>
+    </ol>
+  </li>
+  <li><h4><a id="security_sasl_clientconfig"
+    href="#security_sasl_clientconfig">SASL configuration for Kafka clients</a></h4>
+    SASL authentication is only supported for the new kafka producer and
```
Comment (junrao) [inline docs/security.html]: To make it clear, perhaps change "new" to "new java".

File: docs/security.html (hunk 1 above)
Comment (junrao) [inline docs/security.html]: Yes, second on Magnus question. We should make it clear that username/password is used for broker to initiate a connection to other brokers and user_userName is used for broker to validate all client connections (including those from other brokers).

File: docs/security.html [hunk 3]
```diff
@@ -257,93 +322,191 @@
...
+          <li>Configure SASL port and SASL mechanisms in server.properties as described <a href="#security_sasl_brokerconfig">here</a>.</pre> For example:
+            <pre>    listeners=SASL_SSL://host.name:port
+    security.inter.broker.protocol=SASL_SSL
+    sasl.mechanism.inter.broker.protocol=PLAIN
+    sasl.enabled.mechanisms=PLAIN</pre></li>
+        </ol>
+      </li>
+
+      <li><h5><a id="security_sasl_plain_clientconfig" href="#security_sasl_plain_clientconfig">Configuring Kafka Clients</a></h5>
+        To configure SASL authentication  on the clients:
```
Comment (junrao) [inline docs/security.html]: extra space before on.

File: docs/security.html [hunk 4]
```diff
@@ -257,93 +322,191 @@
...
+          <li>The default implementation of SASL/PLAIN in Kafka specifies usernames and passwords in the JAAS configuration file as shown
+            <a href="#security_sasl_plain_brokerconfig">here</a>. To avoid storing passwords on disk, you can plugin your own implementation of
+            <code>javax.security.auth.spi.LoginModule</code> that provides usernames and passwords from an external source. The login module implementation should
+            provide username as the public credential and password as the private credential of the <code>Subject</code>. The default implementation
+            <code>org.apache.kafka.common.security.plain.PlainLoginModule</code> can be used as an example.</li>
+          <li>In production systems, external authentication servers may implement password authentication. Kafka brokers can be integrated with these servers by adding
+            your own implementation of <code>javax.security.sasl.SaslServer</code>. The default implementation included in Kafka in the package
+            <code>org.apache.kafka.common.security.plain</code> can be used as an example to get started.</li>
```
Comment (junrao) [inline docs/security.html]: Could we describe how the user can plug in a customized PLAIN implementation (e.g., configuring the provider)?

**Comment by junrao** [comment]
@rajinisivaram : Thanks for the patch. Could you also add a description of the latest SASL authentication protocol in protocol.html?

File: docs/security.html (hunk 1 above)
Comment (rajinisivaram) [inline docs/security.html]: Have updated the docs.

File: docs/security.html (hunk 2 above)
Comment (rajinisivaram) [inline docs/security.html]: Done.

File: docs/security.html (hunk 3 above)
Comment (rajinisivaram) [inline docs/security.html]: Removed.

File: docs/security.html (hunk 4 above)
Comment (rajinisivaram) [inline docs/security.html]: Done.

**Comment by rajinisivaram** [comment]
@edenhill @junrao Thank you both for the reviews. I have updated the docs and added a section in protocol.html. Please let me know if more details should be added.

**Comment by junrao** [comment]
@rajinisivaram : Thanks for the patch. LGTM

**Pull request closed without merging**